support/controlplane-component/defaults.go

```diff
@@ -101,62 +101,101 @@ func (c *controlPlaneWorkload[T]) defaultOptions(cpContext ControlPlaneContext,
+		switch {
+		case v.ConfigMap != nil:
+			names.Insert(v.ConfigMap.Name)
+		case v.Projected != nil:
+			for _, source := range v.Projected.Sources {
+				if source.ConfigMap != nil {
+					names.Insert(source.ConfigMap.Name)
+				}
 			}
-			return err
-		}
-
-		for _, value := range secret.Data {
-			hashedData = append(hashedData, util.HashSimple(value))
 		}
 	}
+	return sets.List(names)
+}
 
-	for configMapName := range configMapsNames {
-		configMap := &corev1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      configMapName,
-				Namespace: cpContext.HCP.Namespace,
-			},
-		}
-		if err := cpContext.Client.Get(cpContext, client.ObjectKeyFromObject(configMap), configMap); err != nil {
-			if apierrors.IsNotFound(err) {
-				continue
+func podSecretNames(spec *corev1.PodSpec) []string {
+	names := sets.New[string]()
+	for _, v := range spec.Volumes {
+		switch {
+		case v.Secret != nil:
+			names.Insert(v.Secret.SecretName)
+		case v.Projected != nil:
+			for _, source := range v.Projected.Sources {
+				if source.Secret != nil {
+					names.Insert(source.Secret.Name)
+				}
 			}
-			return err
 		}
+	}
+	return sets.List(names)
+}
 
-		for _, value := range configMap.Data {
-			hashedData = append(hashedData, util.HashSimple(value))
+func fetchResource[T client.Object](ctx context.Context, obj T, namespace string, c client.Client) func(string) (T, error) {
+	return func(name string) (T, error) {
+		resource := obj.DeepCopyObject().(client.Object)
+		resource.SetName(name)
+		resource.SetNamespace(namespace)
+		if err := c.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil && !apierrors.IsNotFound(err) {
+			return obj, err
 		}
+		return resource.(T), nil
+	}
+}
+
+func (c *controlPlaneWorkload[T]) applyWatchedResourcesAnnotation(cpContext ControlPlaneContext, podTemplate *corev1.PodTemplateSpec) error {
+	// remove duplicate entries if any.
+	secretNames := podSecretNames(&podTemplate.Spec)
+	configMapNames := podConfigMapNames(&podTemplate.Spec)
+
+	hashString, err := computeResourceHash(secretNames, configMapNames,
+		fetchResource(cpContext, &corev1.Secret{}, cpContext.HCP.Namespace, cpContext.Client),
+		fetchResource(cpContext, &corev1.ConfigMap{}, cpContext.HCP.Namespace, cpContext.Client))
+	if err != nil {
+		return err
 	}
-	// if not sorted, we could get a different value on each reconciliation loop and cause unneeded rollout.
-	slices.Sort(hashedData)
 
 	if podTemplate.Annotations == nil {
 		podTemplate.Annotations = map[string]string{}
 	}
-	podTemplate.Annotations["component.hypershift.openshift.io/config-hash"] = strings.Join(hashedData, "")
+	podTemplate.Annotations["component.hypershift.openshift.io/config-hash"] = hashString
 	return nil
 }
 
+func computeResourceHash(secretNames, configMapNames []string,
+	fetchSecret func(string) (*corev1.Secret, error),
+	fetchConfigMap func(string) (*corev1.ConfigMap, error),
+) (string, error) {
+	slices.Sort(secretNames)
+	slices.Sort(configMapNames)
```

Comment:
We don't really need to sort here, sorting the hashes at the end `slices.Sort(hashes)` should be sufficient, right?

Reply:
removed